packages/near-api-js/src/contract.ts

```diff
@@ -90,45 +119,92 @@ export class Contract {
     constructor(account: Account, contractId: string, options: ContractMethods) {
         this.account = account;
         this.contractId = contractId;
-        const { viewMethods = [], changeMethods = [] } = options;
-        viewMethods.forEach((methodName) => {
-            Object.defineProperty(this, methodName, {
+        const { viewMethods = [], changeMethods = [], abi: abiRoot } = options;
+
+        let viewMethodsWithAbi = viewMethods.map((name) => ({ name, abi: null as AbiFunction }));
+        let changeMethodsWithAbi = changeMethods.map((name) => ({ name, abi: null }));
+        if (abiRoot) {
+            const abiViewMethods = abiRoot.body.functions
+                .filter((m) => m.kind === AbiFunctionKind.View)
+                .map((m) => ({ name: m.name, abi: m }));
+            const abiChangeMethods = abiRoot.body.functions
+                .filter((methodAbi) => methodAbi.kind === AbiFunctionKind.Call)
+                .map((methodAbi) => ({ name: methodAbi.name, abi: methodAbi }));
+            viewMethodsWithAbi = viewMethodsWithAbi.concat(abiViewMethods);
+            changeMethodsWithAbi = changeMethodsWithAbi.concat(abiChangeMethods);
+        }
+
+        // Strict mode is disabled for now as it complains about unknown formats. We need to
+        // figure out if we want to support a fixed set of formats. `uint32` and `uint64`
+        // are added explicitly just to reduce the amount of warnings as these are very popular
+        // types.
+        const ajv = new Ajv({
+            strictSchema: false,
+            formats: {
+                uint32: true,
+                uint64: true
+            }
+        });
+        addFormats(ajv);
+
+        viewMethodsWithAbi.forEach(({ name, abi }) => {
+            Object.defineProperty(this, name, {
                 writable: false,
                 enumerable: true,
-                value: nameFunction(methodName, async (args: object = {}, options = {}, ...ignored) => {
+                value: nameFunction(name, async (args: object = {}, options = {}, ...ignored) => {
                     if (ignored.length || !(isObject(args) || isUint8Array(args)) || !isObject(options)) {
                         throw new PositionalArgsError();
                     }
+
+                    if (abi) {
+                        if (abi.params && abi.params.serialization_type !== AbiSerializationType.Json) {
```

Comment:
💄 You can use optional chaining here
```js
abi.params?.serialization_type !== AbiSerializationType.Json)
```

Reply:
Wait, I just realized that this is not equivalent to the original line. `abi.params?.serialization_type !== AbiSerializationType.Json` also returns true if `abi.params`/`abi.result` is undefined which is not what we want.

Reply:
🤦 my bad!